coderd/notifications/reports/generator.go

```diff
@@ -0,0 +1,296 @@
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	"github.com/coder/coder/v2/coderd/notifications"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+const (
+	delay = 15 * time.Minute
+)
+
+func NewReportGenerator(ctx context.Context, logger slog.Logger, db database.Store, enqueuer notifications.Enqueuer, clk quartz.Clock) io.Closer {
+	closed := make(chan struct{})
+
+	ctx, cancelFunc := context.WithCancel(ctx)
+	//nolint:gocritic // The system generates periodic reports without direct user input.
+	ctx = dbauthz.AsSystemRestricted(ctx)
+
+	// Start the ticker with the initial delay.
+	ticker := clk.NewTicker(delay)
+	doTick := func(start time.Time) {
+		defer ticker.Reset(delay)
+		// Start a transaction to grab advisory lock, we don't want to run generator jobs at the same time (multiple replicas).
+		if err := db.InTx(func(tx database.Store) error {
+			// Acquire a lock to ensure that only one instance of the generator is running at a time.
+			ok, err := tx.TryAcquireLock(ctx, database.LockIDReportGenerator)
+			if err != nil {
+				return err
+			}
+			if !ok {
+				logger.Debug(ctx, "unable to acquire lock for generating periodic reports, skipping")
+				return nil
+			}
+
+			err = reportFailedWorkspaceBuilds(ctx, logger, db, enqueuer, clk)
+			if err != nil {
+				logger.Debug(ctx, "unable to report failed workspace builds")
+				return err
+			}
+
+			logger.Info(ctx, "report generator finished", slog.F("duration", clk.Since(start)))
+
+			return nil
+		}, nil); err != nil {
+			logger.Error(ctx, "failed to generate reports", slog.Error(err))
+			return
+		}
+	}
+
+	go func() {
+		defer close(closed)
+		defer ticker.Stop()
+		// Force an initial tick.
+		doTick(dbtime.Time(clk.Now()).UTC())
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case tick := <-ticker.C:
+				ticker.Stop()
+				doTick(dbtime.Time(tick).UTC())
+			}
+		}
+	}()
+	return &reportGenerator{
+		cancel: cancelFunc,
+		closed: closed,
+	}
+}
+
+type reportGenerator struct {
+	cancel context.CancelFunc
+	closed chan struct{}
+}
+
+func (i *reportGenerator) Close() error {
+	i.cancel()
+	<-i.closed
+	return nil
+}
+
+const failedWorkspaceBuildsReportFrequencyDays = 7
+
+func reportFailedWorkspaceBuilds(ctx context.Context, logger slog.Logger, db database.Store, enqueuer notifications.Enqueuer, clk quartz.Clock) error {
+	now := clk.Now()
+	since := now.Add(-failedWorkspaceBuildsReportFrequencyDays * 24 * time.Hour)
+
+	statsRows, err := db.GetWorkspaceBuildStatsByTemplates(ctx, dbtime.Time(since).UTC())
+	if err != nil {
+		return xerrors.Errorf("unable to fetch failed workspace builds: %w", err)
+	}
+	sort.Slice(statsRows, func(i, j int) bool {
+		return statsRows[i].TemplateName < statsRows[j].TemplateName
+	})
+
+	reportGeneratedNow := map[uuid.UUID]bool{}
+	for _, stats := range statsRows {
+		var failedBuilds []database.GetFailedWorkspaceBuildsByTemplateIDRow
+		reportData := map[string]any{}
+
+		if stats.FailedBuilds > 0 {
```

Comment:
Can you perhaps invert this and `continue`?
Looks like we do a bunch of work we don't need to if there are no failed builds.